CIPs/cip-granda_mento.md

```diff
@@ -0,0 +1,187 @@
+
+The contract has the following configurable parameters:
+
+1. **`address public approver;`** - Set by Governance. Intended to be a multisig and has the authority to approve exchanges. The signers for the multisig are at the discretion of Governance, but it would comprise of community members or members on behalf of organizations that are aligned with the Celo network/ecosystem.
+2. **`uint256 public exchangeWaitPeriodSeconds;`** - Set by Governance. The minimum amount of time in seconds that must elapse between a proposed exchange being approved and when the exchange can be executed. Should give sufficient time for Governance to veto an approved exchange.
+3. **`FixidityLib.Fraction public spread;`** - Set by Governance. The percent fee imposed upon an exchange execution.
+4. **`mapping(address => ExchangeLimits) stableTokenExchangeLimits`** - Set by Governance via `setStableTokenExchangeLimits`. A mapping of stable token address to a struct containing two `uint256`s, `minExchangeAmount` and `maxExchangeAmount`.
+
+The contract has the following functions:
+
+1. **`function proposeExchange(address stableToken, uint256 sellAmount, bool sellCelo) external returns (uint256)`** - Called by an exchange proposer to propose an exchange.
+   * Callable by anyone.
+   * If `sellCelo` is true, CELO is the asset being sold and `stableToken` is the asset being bought. If `sellCelo` is false, `stableToken` is the asset being sold and CELO is the asset being bought.
+   * Requires the amount of `stableToken` being bought/sold to be within the token-specific range `[stableTokenExchangeLimits[stableToken].minExchangeAmount, stableTokenExchangeLimits[stableToken].maxExchangeAmount]` that is set by Governance via `setStableTokenExchangeLimits` (described later).
+   * Deposits the full amount of the asset being sold into the contract.
+   * Records:
+     * The exchange as in the Proposed state.
+     * A struct with the following info is stored in a mapping with an `id` key:
+       * The stable token address.
+       * The amount of the asset being sold.
+         * Behind the scenes, StableToken keeps track of balances via "units," which do not change with time. However, the value returned by a call to `balanceOf` is the "value," which can change with time due to the inflation/demurrage feature. Because GrandaMento must refund a deposit for a Cancelled exchange proposal, GrandaMento should be sure to refund the correct amount adjusted by any inflation that has been incurred. To account for this, the amount of the StableToken being sold is stored in "units," and the "value" at a given time is calculated using [StableToken.unitsToValue](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L415).
```

Comment:
@yorhodes What do you think of us changing this to `public`, or adding a getter or something?

https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L49

That way instead of doing:
```
units = stableToken.valueToUnits(
  stableToken.balanceOf(...) // this has a `unitsToValue` call in there
)
```
we could just do w/ less gas
```
units = stableToken.balanceOfUnits(...)
```

Reply:
I don't really understand this limitation
Could we make the `sellAmount` denominated in units?